### PR TITLE
BUG fix on some critical problems & data structure improvement

### PR DESCRIPTION
**Pull request by WEIFENG-NUSCEG**
BUGS FIX: 
1. a bug was found inside the **StorageManager** causing the storage saving system to save both patient Id and task Id as the same. (Fixed)
2.a bug was found inside the **StorageManager**  causing the storage loading system not loading the unique Id, instead it will regenerate the unique id every time. (Fixed)
3.one exception is not properly handled in the **DeleteAssignedTasks** (Fixed)
4.a bug was found in the **FindAssignedTasksCommand** a patient id is passed into a **TaskManager** (FIXED)
5.Same assigned task can be assigned with the same time period or deadline (Fixed)

New Bugs spotted and has been fixed:
1. **UndoParser** is not catching the command properly, will be fixed soon. (Fixed)
2.The **MementoManger** is previously pop, and add on different side of the **LinkedList**, thus use command undo will actually retrieve the earliest step instead of the last step (Fixed)
3.When we are trying to load a **ArrayList of AssignedTask** to the **AssignedTaskManager**, set the maxId inside our **AssignedTaskManager**, it retrieves the last element from the **Multimap**. However this **Multimap** whose key is a patient ID instead of a Unique ID, thus the last element will not return the actual current **MaxId**(Fixed)

**Data Structure of the AssignedTaskManager has been updated.**
The efficiency of the following operations will be much faster:
1.**public void deletePatientTaskByUniqueId(int uid)** 
previous efficiency will take O(mxn), now it has been reduced to O(n)
(m is the number of patients, n is the number of tasks a patient has)

1.finding if such unique id exists

### CODE REVIEW

File: src/main/java/duke/models/assignedtasks/AssignedTaskWithPeriod.java [hunk 1]
```diff
@@ -59,6 +59,35 @@ public AssignedTaskWithPeriod(int pid, int tid, boolean isDone,
...
+     * @param isDone       .
+     * @param isRecurrsive .
+     * @param stime        .
+     * @param etime        .
+     * @param type         .
+     * @param uniqueId     .
+     */
+    public AssignedTaskWithPeriod(int pid, int tid, boolean isDone,
+                                  boolean isRecurrsive, String stime, String etime, String type, int uniqueId)
+            throws DukeException {
```
Comment (qjie7) [inline src/main/java/duke/models/assignedtasks/AssignedTaskWithPeriod.java]: Perhaps separate the parameters to 3 rows ?

Comment (WEIFENG-NUSCEG) [reply src/main/java/duke/models/assignedtasks/AssignedTaskWithPeriod.java]: sure

File: src/main/java/duke/models/assignedtasks/AssignedTaskManager.java [hunk 2]
```diff
@@ -29,6 +31,7 @@ public AssignedTaskManager(ArrayList<AssignedTask> assignedTasks) {
                 assignedTask.setUuid(maxId);
```
Comment (qjie7) [inline src/main/java/duke/models/assignedtasks/AssignedTaskManager.java]: There are some problem in AssignedTaskManager’s Data structure. When you are trying to load a arraylist of AssignedTask into the AssignedTaskManager, you set the current maxId(unique Is of the task) as the last element in the Multimaps’s unique Id+1 , however this multimap’s key is not this unique Id but it’s patient Id. Therefore the last element’s unique id is not the actual maxId.

Comment (HUANGXUANKUN) [reply src/main/java/duke/models/assignedtasks/AssignedTaskManager.java]: @qjie7 Nice catch. It makes sense. 
@WEIFENG-NUSCEG During loading/saving, the records are listed in according to the patientID. Thus your maxID may not be your maxUuid.

Comment (WEIFENG-NUSCEG) [reply src/main/java/duke/models/assignedtasks/AssignedTaskManager.java]: Hi thanks for the catch, the problem has been solved in the new commit.

**Pull request merged**